enterprise/server/util/redisutil/redisutil.go

```diff
@@ -96,3 +97,141 @@ func (r *redlock) Unlock(ctx context.Context) error {
 	_, err := r.rmu.UnlockContext(ctx)
 	return err
 }
+
+// CommandBuffer buffers and aggregates Redis commands in-memory and allows
+// flushing the aggregate results in batch. This is useful for reducing the load
+// placed on Redis in cases where a high volume of commands are operating on
+// a relatively small subset of Redis keys (for example, counter values that
+// are incremented several times over a short time period).
+//
+// When using this buffer, the caller is responsible for explicitly flushing
+// the data. In most cases, flushing should be done at regular intervals, as well
+// as after writing the last command to the buffer, if applicable.
+//
+// There may be a non-negligible delay between the time that the data is written
+// to the buffer and the time that data is flushed, so callers may need extra
+// synchronization in order to ensure that buffers are flushed (on all nodes,
+// if applicable), before relying on the values of any keys updated via this
+// buffer.
+//
+// It is safe for concurrent access.
+type CommandBuffer struct {
+	rdb *redis.Client
+
+	mu sync.Mutex // protects all fields below
+	// Buffer for INCRBY commands.
+	incr map[string]int64
+	// Buffer for HINCRBY commands.
+	hincr map[string]map[string]int64
+	// Buffer for SADD commands.
+	sadd map[string]map[interface{}]struct{}
+	// Buffer for EXPIRE commands.
+	expire map[string]time.Duration
+}
+
+// NewCommandBuffer creates a buffer.
+func NewCommandBuffer(rdb *redis.Client) *CommandBuffer {
+	c := &CommandBuffer{rdb: rdb}
+	c.init()
+	return c
+}
+
+func (c *CommandBuffer) init() {
+	c.incr = map[string]int64{}
+	c.hincr = map[string]map[string]int64{}
+	c.sadd = map[string]map[interface{}]struct{}{}
+	c.expire = map[string]time.Duration{}
+}
+
+// IncrBy adds an INCRBY operation to the buffer.
+func (c *CommandBuffer) IncrBy(key string, increment int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.incr[key] += increment
+}
+
+// HIncrBy adds an HINCRBY operation to the buffer.
+func (c *CommandBuffer) HIncrBy(key, field string, increment int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	h, ok := c.hincr[key]
+	if !ok {
+		h = map[string]int64{}
+		c.hincr[key] = h
+	}
+	h[field] += increment
+}
+
+// SAdd adds an SADD operation to the buffer. All buffered members of the set
+// will be added in a single SADD command.
+func (c *CommandBuffer) SAdd(key string, members ...interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	s, ok := c.sadd[key]
+	if !ok {
+		s = map[interface{}]struct{}{}
+		c.sadd[key] = s
+	}
+	for _, m := range members {
+		s[m] = struct{}{}
+	}
+}
+
+// Expire adds an EXPIRE operation to the buffer, overwriting any previous
+// expiry currently buffered for the given key. Note that the expiry duration
+// will apply only from the time that the buffer is flushed. This implies that
+// the expiration does not affect any buffered keys.
```

Comment:
"This implies that the expiration does not affect any buffered keys."

Is this true? Since you apply EXPIRE commands last, could a caller of this class do a hincrby, then expire, on the same key and have that work?

i'm just suggesting you remove this part of the comment, unless it's true and i'm misunderstanding!

Reply:
Reworded a bit -- trying to convey that the buffer does not actively apply the expirations, and because expirations overwrite the previous ones, some sequences of commands won't behave the same as if they were issued directly to Redis.